aten/src/ATen/native/Copy.cpp

```diff
@@ -320,6 +320,10 @@ void copy_ignoring_overlaps(const TensorBase &dst, const TensorBase &src) {
   copy_stub(iter.device_type(), iter, /*non_blocking=*/false);
 }
 
+void _propagate_xla_data(const Tensor& input, const Tensor& output) {
+  TORCH_INTERNAL_ASSERT(false, "This op should only be called by XLA")
```

Comment:
@bdhirsh It looks like I cannot use `false` here as XLA's Dynamo integration will dry run this op...

Reply:
hmmmm..... that seems like a bug (since the idea behind this op is that morally, none of it should "need" to exist in the pytorch core codebase. We're really just adding a stub, that XLA can override later).

Are you able to figure out why this kernel is getting hit? I can give some pointers:


(1) This kernel is registered to the `CompositeImplicitAutograd` dispatch key

(1) The XLA kernel in pytorch/xla is registered to the `XLA` dispatch key.

You can sanity check what the disaptcher is seeing by running this:

```
import torch
import torch_xla # register your xla impl to the dispatcher
print(torch._C._dispatch_dump("aten::_propagate_xla_data"))
print(torch._C._dispatch_dump_table("aten::_propagate_xla_data"))
```

Reply:
Let me follow up.

Reply:
It looks like that I forget to update the XLA pin. Let's see if the latest commit can pass the CI. Good catch!

Reply:
Well, actually the error is real. The test is currently disabled on the XLA branch and that's why I didn't catch that last time:
```
======================================================================
ERROR: test_resnet18 (__main__.DynamoTrainingBasicTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/workspaces/work/pytorch/torch/_dynamo/output_graph.py", line 705, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.fake_example_inputs())
  File "/workspaces/work/pytorch/torch/_dynamo/debug_utils.py", line 1055, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/workspaces/work/pytorch/torch/_dynamo/backends/common.py", line 48, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
  File "/workspaces/work/pytorch/torch/_functorch/aot_autograd.py", line 2798, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
  File "/workspaces/work/pytorch/torch/_dynamo/utils.py", line 163, in time_wrapper
    r = func(*args, **kwargs)
  File "/workspaces/work/pytorch/torch/_functorch/aot_autograd.py", line 2491, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config)
  File "/workspaces/work/pytorch/torch/_functorch/aot_autograd.py", line 1802, in aot_wrapper_dedupe
    compiled_fn = compiler_fn(wrapped_flat_fn, deduped_flat_args, aot_config)
  File "/workspaces/work/pytorch/torch/_functorch/aot_autograd.py", line 2033, in aot_dispatch_autograd
    _fw_metadata, out = run_functionalized_fw_and_collect_metadata(
  File "/workspaces/work/pytorch/torch/_functorch/aot_autograd.py", line 606, in inner
    flat_f_outs = f(*flat_f_args)
  File "/workspaces/work/pytorch/torch/_functorch/aot_autograd.py", line 1800, in wrapped_flat_fn
    return flat_fn(*add_dupe_args(args))
  File "/workspaces/work/pytorch/torch/_functorch/aot_autograd.py", line 2769, in functional_call
    out = Interpreter(mod).run(*args[params_len:], **kwargs)
  File "/workspaces/work/pytorch/torch/fx/interpreter.py", line 136, in run
    self.env[node] = self.run_node(node)
  File "/workspaces/work/pytorch/torch/fx/interpreter.py", line 177, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/workspaces/work/pytorch/torch/fx/interpreter.py", line 294, in call_module
    return submod(*args, **kwargs)
  File "/workspaces/work/pytorch/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/workspaces/work/pytorch/torch/nn/modules/batchnorm.py", line 151, in forward
    self.num_batches_tracked.add_(1)  # type: ignore[has-type]
RuntimeError: false INTERNAL ASSERT FAILED at "/workspaces/work/pytorch/aten/src/ATen/native/Copy.cpp":324, please report a bug to PyTorch. This op should only be called by XLA

While executing %model_bn1 : [#users=1] = call_module[target=model_bn1](args = (%model_conv1,), kwargs = {})
Original traceback:
  File "/workspaces/work/vision/torchvision/models/resnet.py", line 269, in _forward_impl
    x = self.bn1(x)
 |   File "/workspaces/work/vision/torchvision/models/resnet.py", line 285, in forward
    return self._forward_impl(x)
 |   File "test/dynamo/test_dynamo.py", line 90, in train_model
    pred = model(data)
```

Reply:
Here is the output from the command:
```
>>> import torch
>>> import torch_xla
>>> print(torch._C._dispatch_dump("aten::_propagate_xla_data"))
name: aten::_propagate_xla_data
schema: aten::_propagate_xla_data(Tensor input, Tensor output) -> ()
debug: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterSchema.cpp:6
alias analysis kind: FROM_SCHEMA
Functionalize: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterFunctionalization_1.cpp:24276 :: (Tensor _0, Tensor _1) -> () [ boxed unboxed ]
Tracer: registered at /workspaces/work/pytorch/torch/csrc/autograd/generated/TraceType_2.cpp:16872 :: (Tensor _0, Tensor _1) -> () [ boxed unboxed ]
XLA: registered at /workspaces/work/pytorch/xla/torch_xla/csrc/generated/RegisterXLA.cpp:4644 :: (Tensor _0, Tensor _1) -> () [ boxed unboxed ]
CompositeImplicitAutograd[alias]: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 :: (Tensor _0, Tensor _1) -> () [ boxed unboxed ]

>>> print(torch._C._dispatch_dump_table("aten::_propagate_xla_data"))
Undefined: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
CPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
CUDA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
HIP: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
XLA: registered at /workspaces/work/pytorch/xla/torch_xla/csrc/generated/RegisterXLA.cpp:4644 [kernel]
MPS: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
IPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
XPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
HPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
VE: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
Lazy: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
Meta: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
MTIA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
PrivateUse1: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
PrivateUse2: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
PrivateUse3: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
FPGA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
ORT: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
Vulkan: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
Metal: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedCPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedCUDA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedHIP: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedXLA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedMPS: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedIPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedXPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedHPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedVE: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedLazy: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedMeta: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedMTIA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedPrivateUse1: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedPrivateUse2: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
QuantizedPrivateUse3: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
CustomRNGKeyId: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
MkldnnCPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseCPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseCUDA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseHIP: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseXLA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseMPS: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseIPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseXPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseHPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseVE: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseLazy: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseMeta: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseMTIA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparsePrivateUse1: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparsePrivateUse2: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparsePrivateUse3: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseCsrCPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
SparseCsrCUDA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorCPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorCUDA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorHIP: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorXLA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorMPS: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorIPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorXPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorHPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorVE: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorLazy: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorMeta: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorMTIA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorPrivateUse1: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorPrivateUse2: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
NestedTensorPrivateUse3: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
BackendSelect: fallthrough registered at /workspaces/work/pytorch/aten/src/ATen/core/BackendSelectFallbackKernel.cpp:3 [backend fallback]
Python: registered at /workspaces/work/pytorch/aten/src/ATen/core/PythonFallbackKernel.cpp:144 [backend fallback]
FuncTorchDynamicLayerBackMode: registered at /workspaces/work/pytorch/aten/src/ATen/functorch/DynamicLayer.cpp:491 [backend fallback]
Functionalize: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterFunctionalization_1.cpp:24276 [kernel]
Named: registered at /workspaces/work/pytorch/aten/src/ATen/core/NamedRegistrations.cpp:7 [backend fallback]
Conjugate: registered at /workspaces/work/pytorch/aten/src/ATen/ConjugateFallback.cpp:17 [backend fallback]
Negative: registered at /workspaces/work/pytorch/aten/src/ATen/native/NegateFallback.cpp:19 [backend fallback]
ZeroTensor: registered at /workspaces/work/pytorch/aten/src/ATen/ZeroTensorFallback.cpp:86 [backend fallback]
ADInplaceOrView: fallthrough registered at /workspaces/work/pytorch/aten/src/ATen/core/VariableFallbackKernel.cpp:63 [backend fallback]
AutogradOther: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradCPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradCUDA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradHIP: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradXLA: fallthrough registered at /workspaces/work/pytorch/aten/src/ATen/core/VariableFallbackKernel.cpp:46 [backend fallback]
AutogradMPS: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradIPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradXPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradHPU: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradVE: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradLazy: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradMeta: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradMTIA: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradPrivateUse1: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradPrivateUse2: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradPrivateUse3: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
AutogradNestedTensor: registered at /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:7156 [math kernel]
Tracer: registered at /workspaces/work/pytorch/torch/csrc/autograd/generated/TraceType_2.cpp:16872 [kernel]
AutocastCPU: fallthrough registered at /workspaces/work/pytorch/aten/src/ATen/autocast_mode.cpp:487 [backend fallback]
AutocastCUDA: fallthrough registered at /workspaces/work/pytorch/aten/src/ATen/autocast_mode.cpp:354 [backend fallback]
FuncTorchBatched: registered at /workspaces/work/pytorch/aten/src/ATen/functorch/LegacyBatchingRegistrations.cpp:815 [backend fallback]
FuncTorchVmapMode: fallthrough registered at /workspaces/work/pytorch/aten/src/ATen/functorch/VmapModeRegistrations.cpp:28 [backend fallback]
Batched: registered at /workspaces/work/pytorch/aten/src/ATen/LegacyBatchingRegistrations.cpp:1073 [backend fallback]
VmapMode: fallthrough registered at /workspaces/work/pytorch/aten/src/ATen/VmapModeRegistrations.cpp:33 [backend fallback]
FuncTorchGradWrapper: registered at /workspaces/work/pytorch/aten/src/ATen/functorch/TensorWrapper.cpp:210 [backend fallback]
PythonTLSSnapshot: registered at /workspaces/work/pytorch/aten/src/ATen/core/PythonFallbackKernel.cpp:152 [backend fallback]
FuncTorchDynamicLayerFrontMode: registered at /workspaces/work/pytorch/aten/src/ATen/functorch/DynamicLayer.cpp:487 [backend fallback]
PythonDispatcher: registered at /workspaces/work/pytorch/aten/src/ATen/core/PythonFallbackKernel.cpp:148 [backend fallback]
```

Reply:
Maybe we need to register more than 'XLA' key for Dynamo's AOTAutograd?

Reply:
Hmm.... I'm not sure. One guess is that it might be because we're now e.g. trying to call the op somewhere with meta tensors during tracing, which would hit the `Meta` dispatch key (and hit the composite kernel).

Ok - I think keeping the original assert that you that you had (asserting the input device is XLA instead of blank asserting false is reasonable enough. If you want to change it back, happy to land that. Thanks for checking with me!

Reply:
Reverting the change back.